arbor/mc_cell_group.cpp

```diff
@@ -79,6 +80,245 @@ void mc_cell_group::set_binning_policy(binning_kind policy, time_type bin_interv
     binners_.resize(gids_.size(), event_binner(policy, bin_interval));
 }
 
+// Probe-type specific sample data marshalling.
+
+struct sampler_call_info {
+    sampler_function sampler;
+    cell_member_type probe_id;
+    probe_tag tag;
+
+    // Offsets are into lowered cell sample time and event arrays.
+    sample_size_type begin_offset;
+    sample_size_type end_offset;
+};
+
+// Working space for computing and collating data for samplers.
+using fvm_probe_scratch = std::tuple<std::vector<double>, std::vector<cell_sample_range>>;
+
+template <typename VoidFn, typename... A>
+void tuple_foreach(VoidFn&& f, std::tuple<A...>& t) {
+    (void)(int []){(f(std::get<A>(t)), 0)...};
+}
+
+void reserve_scratch(fvm_probe_scratch& scratch, std::size_t n) {
+    tuple_foreach([n](auto& v) { v.reserve(n); }, scratch);
```

Comment:
Given that `tuple_foreach` is used only once, right below where it is defined, I think it would be easier to just say:
```C++
scratch.get<0>().reserve(n);
scratch.get<1>().reserve(n);
```

Looking at `fvm_probe_scratch` and how it is used, it looks like it could be a simple struct. Is there reason it is a `tuple`?

Reply:
At one point, there were more vectors in scratch, and a struct was getting unwieldy and copy/pastey.

With just the two, a struct is not bad. But the tuple approach has the benefit of being easy to generalize. It doesn't strike me as being particularly obscure as a tuple, but that's probably a matter of taste.